File: 402-remove-k-digits/remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        n = len(num)
        st = []
        for i in num:
            while st and k>0 and st[-1] > i:
                st.pop()
                k-=1
            st.append(i)
        while k>0:
            st.pop()
            k-=1
        res = ""
        while st:
            res += st.pop()
        while len(res)!=0 and res[-1] == '0':
            res = res[:-1]
        res = res[::-1]
        return res if res else "0"
        

File: 402-remove-k-digits/remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        stack<char>st;
        int n = num.size();
        for(int i = 0; i<n; i++){
            while(!st.empty() && k > 0 && (st.top() > num[i])){
                st.pop();
                k--;
            }
            st.push(num[i]);
        }
        while(k>0){ //edge case for strictly increasing numbers
            st.pop();
            k--;
        }
        if(st.empty()) return "0"; //edge case
        string res = "";
        while(!st.empty()){
            res+=st.top();
            st.pop();
        }
        while(res.size()!=0 && res.back() == '0'){ //edge case for leading 0s
            res.pop_back();
        }
        reverse(res.begin(), res.end());
        return res.empty() ? "0" : res;
    }
};